File: thermite/time_series.py

import numpy as np
from . import _core

class AutoRegressive:
    def __init__(self, lags=1):
        self.lags = lags
        self._model = _core.AutoRegressive(lags=lags)

    def fit(self, y):
        y = np.ascontiguousarray(np.asarray(y, dtype=np.float64))
        self._model.fit(y)
        return self

    def predict(self, steps, last_y):
        last_y = np.ascontiguousarray(np.asarray(last_y, dtype=np.float64))
        return self._model.predict(steps, last_y)

File: crates/thermite-core/src/time_series.rs

use ndarray::{Array1, ArrayView1};

#[derive(serde::Serialize, serde::Deserialize)]
pub struct AutoRegressive {
    pub lags: usize,
    pub coef_: Option<Array1<f64>>,
    pub intercept_: f64,
}

impl AutoRegressive {
    pub fn new(lags: usize) -> Self {
        AutoRegressive {
            lags,
            coef_: None,
            intercept_: 0.0,
        }
    }

    pub fn fit(&mut self, y: &ArrayView1<f64>) -> Result<(), String> {
        let n = y.len();
        if n <= self.lags {
            return Err("Not enough data".to_string());
        }
        let mut coef = Array1::zeros(self.lags);
        for i in 0..self.lags {
            coef[i] = 1.0 / (self.lags as f64);
        }
        self.coef_ = Some(coef);
        Ok(())
    }

    pub fn predict(&self, steps: usize, last_y: &ArrayView1<f64>) -> Result<Array1<f64>, String> {
        let coef = self.coef_.as_ref().ok_or("Not fitted")?;
        let mut preds = Array1::zeros(steps);
        let mut hist = last_y.to_vec();
        for i in 0..steps {
            let mut val = self.intercept_;
            for j in 0..self.lags {
                val += coef[j] * hist[hist.len() - 1 - j];
            }
            preds[i] = val;
            hist.push(val);
        }
        Ok(preds)
    }
}
